### PR TITLE
Ensure inner lock drop before handle

### PR DESCRIPTION
**Pull request by treere**
Hi!

When one of the guardian struct is the only one keeping the handler alive, during drop the stack/heap is corrupted.

```rust
pub struct Example {
    _handle: Arc<Mutex<i8>>,
    inner: MutexGuard<'static, i8>,
}
```

If `_handle` is dropped before `inner`, when `inner` is dropped it tries to unlock `_handler` but it has already been dropped.

The drop field order in Rust is not fixed so I used an `Option` to force the right order.

Have a nice day!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -37,7 +37,7 @@ use std::ops::DerefMut;
 /// implementations.
 pub struct ArcRwLockReadGuardian<T: 'static> {
     _handle: sync::Arc<sync::RwLock<T>>,
-    inner: sync::RwLockReadGuard<'static, T>,
+    inner: Option<sync::RwLockReadGuard<'static, T>>,
```
Comment (jonhoo) [inline src/lib.rs:43]: Can you also add to the comment above ("ATTENTION READERS") along the lines of:
> Each structure holds the guard in an `Option` to ensure that we drop the guard _before_ we drop the handle, as dropping the guard will access the handle.

File: src/lib.rs [hunk 2]
```diff
@@ -99,42 +99,42 @@ pub struct RcMutexGuardian<T: 'static> {
 impl<T> Deref for ArcRwLockReadGuardian<T> {
     type Target = T;
     fn deref(&self) -> &Self::Target {
-        &*self.inner
+        self.inner.as_ref().unwrap()
```
Comment (jonhoo) [inline src/lib.rs]: Let's make all of these
```suggestion
        self.inner.as_ref().expect("inner is only None in drop")
```

**Review by jonhoo — CHANGES_REQUESTED** [review]
Good catch! Just some small nits!

**Comment by treere** [comment]
Thanks for the review! I have applied the changes

File: src/lib.rs [hunk 3]
```diff
@@ -16,11 +16,14 @@
...
 //! method, so for `RwLock`s, the same rule applies for when a lock may be poisioned.
 
-use std::rc;
-use std::sync;
 use std::ops::Deref;
 use std::ops::DerefMut;
+use std::rc;
+use std::sync;
 
+// Each structure holds the guard in an Option to ensure that we drop the guard before we drop the
```
Comment (jonhoo) [inline src/lib.rs]: Ah, sorry, I phrased my comment poorly. I meant for this to be added _to_ the ATTENTION READERS section, not above it :)

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (treere) [reply src/lib.rs]: My bad! I followed the comment too literally :)

**Review by jonhoo — APPROVED** [review]
Thanks!

**Comment by jonhoo** [comment]
Published as 1.0.2!

**Pull request merged**